purchases/src/main/kotlin/com/revenuecat/purchases/CustomerInfoRetriever.kt

```diff
@@ -0,0 +1,141 @@
+import com.revenuecat.purchases.common.Backend
+import com.revenuecat.purchases.common.LogIntent
+import com.revenuecat.purchases.common.caching.DeviceCache
+import com.revenuecat.purchases.common.debugLog
+import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.common.log
+import com.revenuecat.purchases.interfaces.ReceiveCustomerInfoCallback
+import com.revenuecat.purchases.strings.CustomerInfoStrings
+
+internal class CustomerInfoRetriever(
+    private val deviceCache: DeviceCache,
+    private val backend: Backend,
+    private val handler: Handler = Handler(Looper.getMainLooper())
+) {
+
+    fun retrieveCustomerInfo(
+        appUserID: String,
+        fetchPolicy: CacheFetchPolicy,
+        appInBackground: Boolean,
+        cacheCustomerInfoAfterFetchCallback: (CustomerInfo) -> Unit,
+        callback: ReceiveCustomerInfoCallback? = null
+    ) {
+        debugLog(CustomerInfoStrings.RETRIEVING_CUSTOMER_INFO.format(fetchPolicy))
+        when (fetchPolicy) {
+            CacheFetchPolicy.CACHE_ONLY -> getCustomerInfoCacheOnly(appUserID, callback)
+            CacheFetchPolicy.FETCH_CURRENT -> getCustomerInfoFetchOnly(
+                appUserID,
+                appInBackground,
+                cacheCustomerInfoAfterFetchCallback,
+                callback
+            )
+            CacheFetchPolicy.CACHED_OR_FETCHED -> getCustomerInfoCachedOrFetched(
+                appUserID,
+                appInBackground,
+                cacheCustomerInfoAfterFetchCallback,
+                callback
+            )
+            CacheFetchPolicy.NOT_STALE_CACHED_OR_CURRENT -> getCustomerInfoNotStaledCachedOrFetched(
+                appUserID,
+                appInBackground,
+                cacheCustomerInfoAfterFetchCallback,
+                callback
+            )
+        }
+    }
+
+    private fun getCustomerInfoCacheOnly(
+        appUserID: String,
+        callback: ReceiveCustomerInfoCallback? = null
+    ) {
+        if (callback == null) return
+        val cachedCustomerInfo = deviceCache.getCachedCustomerInfo(appUserID)
+        if (cachedCustomerInfo != null) {
+            log(LogIntent.DEBUG, CustomerInfoStrings.VENDING_CACHE)
+            dispatch { callback.onReceived(cachedCustomerInfo) }
+        } else {
+            val error = PurchasesError(PurchasesErrorCode.CustomerInfoNotFoundInCache)
+            errorLog(error)
+            dispatch { callback.onError(error) }
+        }
+    }
+
+    private fun getCustomerInfoFetchOnly(
+        appUserID: String,
+        appInBackground: Boolean,
+        cacheCustomerInfoAfterFetchCallback: (CustomerInfo) -> Unit,
+        callback: ReceiveCustomerInfoCallback? = null
+    ) {
+        deviceCache.setCustomerInfoCacheTimestampToNow(appUserID)
+        backend.getCustomerInfo(
+            appUserID,
+            appInBackground,
+            { info ->
+                log(LogIntent.RC_SUCCESS, CustomerInfoStrings.CUSTOMERINFO_UPDATED_FROM_NETWORK)
```

Comment:
I moved this log from where it was before to here. Before, it would log success before the request succeeded. I think this is a better place but lmk if you have any concerns